cpp/src/io/parquet/experimental/dictionary_page_filter.cu

```diff
@@ -46,11 +48,1538 @@ using parquet::detail::ColumnChunkDesc;
+    std::vector<ast::generic_scalar_device_view> host_scalars;
+    host_scalars.reserve(total_num_literals);
+    std::for_each(literals.begin(), literals.end(), [&](auto const& literal) {
+      host_scalars.push_back(literal->get_value());
+    });
+    // Device vector of all scalars device views
+    auto const scalars = cudf::detail::make_device_uvector_async(host_scalars, stream, default_mr);
+
+    // Device vector of all operators
+    auto const d_operators = cudf::detail::make_device_uvector_async(operators, stream, default_mr);
+
+    // Device buffers to store the dictionary membership results for all predicates
+    std::vector<rmm::device_buffer> results_buffers(total_num_literals);
+    // Host vector of pointers to the result buffers
+    auto host_results_ptrs = cudf::detail::make_host_vector<bool*>(total_num_literals, stream);
+    std::for_each(
+      thrust::counting_iterator(0), thrust::counting_iterator(total_num_literals), [&](auto i) {
+        // Allocate the results buffer using the user-provided memory resource (output memory)
+        results_buffers[i]   = rmm::device_buffer(total_row_groups, stream, mr);
+        host_results_ptrs[i] = static_cast<bool*>(results_buffers[i].data());
+      });
+    if constexpr (not cuda::std::is_same_v<T, cudf::string_view>) {
+      // Decode fixed width dictionaries and insert them to cuco hash sets, one dictionary per
+      // thread block
+      build_fixed_width_dictionaries<T>
+        <<<total_row_groups, DECODE_BLOCK_SIZE, 0, stream.value()>>>(pages.device_begin(),
+                                                                     chunks.device_begin(),
+                                                                     decoded_data,
+                                                                     set_storage.data(),
+                                                                     set_offsets.data(),
+                                                                     value_offsets.data(),
+                                                                     physical_type,
+                                                                     num_dictionary_columns,
+                                                                     dictionary_col_idx,
+                                                                     error_code.data());
+
+    } else {
+      // Check if the decode block size is a multiple of the warp size
+      static_assert(DECODE_BLOCK_SIZE % cudf::detail::warp_size == 0,
+                    "DECODE_BLOCK_SIZE must be a multiple of warp_size");
+      // Check if the physical type is a string
+      CUDF_EXPECTS(physical_type == parquet::Type::BYTE_ARRAY, "Unsupported physical type");
+
+      // Compute an optimal thread block size for the query kernel so that we can decode all string
+      // dictionary pages, one per thread
+      auto block_size = std::max<cudf::size_type>(cudf::detail::warp_size, total_row_groups);
```

Comment:
Use an optimal block size between [32, 128] to decode string dictionaries (one per thread) and build cuco hash maps